### PR TITLE
Ports /vg/ parallax

### PR DESCRIPTION
**Pull request by monster860**
:cl: monster860, clusterfack, and DeityLink
rscadd: Adds space parallax
/:cl:

Ports d3athrow/vgstation13#10444

In addition to what /vg/ has, this also has a thing that prevents parallax from showing in places that are obscured by opaque objects, and places that are dark.

And also, I decided to add another feature: Parallax smoothing.

[https://gfycat.com/ImmaterialKindheartedKingsnake](https://gfycat.com/ImmaterialKindheartedKingsnake)

### CODE REVIEW

File: SQL/paradise_schema_prefixed.sql [hunk 1]
```diff
@@ -249,6 +249,9 @@ CREATE TABLE `SS13_player` (
   `nanoui_fancy` smallint(4) DEFAULT '1',
   `show_ghostitem_attack` smallint(4) DEFAULT '1',
   `lastchangelog` varchar(32) NOT NULL,
+  'space_parallax' smallint(4) DEFAULT '1',
```
Comment (tigercat2000) [inline SQL/paradise_schema_prefixed.sql]: this is missing in paradise_schema.sql

Comment (CrAzYPiLoT-SS13) [inline SQL/paradise_schema_prefixed.sql]: Shouldn't this be backticks instead of normal ticks too?

Comment (monster860) [inline SQL/paradise_schema_prefixed.sql]: Fixed. Also, you failed your Turing test.

File: SQL/paradise_schema.sql [hunk 2]
```diff
@@ -249,6 +249,9 @@ CREATE TABLE `player` (
   `nanoui_fancy` smallint(4) DEFAULT '1',
   `show_ghostitem_attack` smallint(4) DEFAULT '1',
   `lastchangelog` varchar(32) NOT NULL,
+  `space_parallax` smallint(4) DEFAULT '1',
+  `space_dust` smallint(4) DEFAULT '1',
+  `parallax_speed` smallint(4) DEFAULT '2',
```
Comment (PJB3005) [inline SQL/paradise_schema.sql]: Make this a float, there's currently an issue on /vg/ due to this being an integer (decimals don't save)

**Comment by Norgad** [comment]
Yes please!
❤❤👍

**Comment by KasparoVy** [comment]
Wicked stuff with the backport, haha. Cool to see the team effort this feature's become. Really looking forward to it.

**Comment by Fox-McCloud** [comment]
Don't forget to use the istypeless for loops, where applicable.

**Comment by monster860** [comment]
@Fox-McCloud What exactly is that

File: SQL/paradise_schema.sql (hunk 2 above)
Comment (monster860) [inline SQL/paradise_schema.sql]: What is the type called in sql

**Comment by VampyrBytes** [comment]
@monster860 He means don't cast the type in the loop declaration where you are sure there's only one type in the list as this triggers an istype check on everything in the list
So

`for(var/mob/M in mobList)`

becomes
```
for(var/m in mobList)
    var/mob/M = m
```

File: code/_onclick/hud/parallax.dm [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+	0, 0, 0, 1)
+
+/obj/screen/plane_master/parallax_dustmaster
+	plane = PLANE_SPACE_DUST
+
+/datum/hud/proc/update_parallax()
+	var/client/C = mymob.client
+	if(!parallax_initialized || C.updating_parallax) return
+
+	for(var/turf/T in range(get_turf(C.eye),C.view))
```
Comment (VampyrBytes) [inline code/_onclick/hud/parallax.dm]: This should be `for(var/turf/space/T in range(get_turf(C.eye), C.view)) - at the moment you're doing an istype() on everything in the range twice... once to see if it's a `/turf` and then again to see if it's a `turf/space`

File: code/_onclick/hud/parallax.dm [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+		for(var/obj/screen/parallax/bgobj in parallax_icon)
+			var/obj/screen/parallax/parallax_layer = new /obj/screen/parallax
+			parallax_layer.appearance = bgobj.appearance
+			parallax_layer.base_offset_x = bgobj.base_offset_x
+			parallax_layer.base_offset_y = bgobj.base_offset_y
+			parallax_layer.parallax_speed = bgobj.parallax_speed
+			C.parallax += parallax_layer
+
+	var/parallax_loaded = 0
+	for(var/obj/screen/S in C.screen)
```
Comment (Fox-McCloud) [inline code/_onclick/hud/parallax.dm]: similar to what @VampyrBytes said, above.

**Comment by PJB3005** [comment]
> Don't forget to use the istypeless for loops, where applicable.

Please, it's useless.

**Comment by PJB3005** [comment]
https://github.com/d3athrow/vgstation13/pull/10444#discussion_r66351927

<sup><sup><sup>I'll stop spamming comments now</sup></sup></sup>

File: code/_onclick/hud/parallax.dm (hunk 3 above)
Comment (PJB3005) [inline code/_onclick/hud/parallax.dm]: It should also be using a `trange()`

**Comment by Fox-McCloud** [comment]
Looks like I'll have to bring @RemieRichards in here to correct you @PJB3005 since she has empirical evidence to demonstrate otherwise. As does @Krausus

**Comment by PJB3005** [comment]
@Fox-McCloud bring. It. On.

**Comment by Spacemanspark** [comment]
Server wars 2: the coder wars

**Comment by monster860** [comment]
Coming soon: Smooth parallax.

**Comment by Krausus** [comment]
... huh? I'm pretty sure the times I've intentionally used typeless loops were when I explicitly wanted to catch things that weren't the right type, not for the sake of optimization.

Now, it wouldn't surprise me if there *is* a cost savings to skipping the implicit type checking. It also wouldn't surprise me if it's one of those optimizations that only become meaningful when you're dealing with tens or hundreds of thousands of things at a time.

@PJB3005's mention of premature optimization is on point. There are tons of things we *could* do to make code faster, but don't, because many optimizations cause more harm (to code readability and maintainability) than good. Optimizing is something you should do *after* you know what code is running slow, not *before.*

**Comment by monster860** [comment]
Parallax smoothing!

[https://gfycat.com/ImmaterialKindheartedKingsnake](https://gfycat.com/ImmaterialKindheartedKingsnake)

**Comment by Fox-McCloud** [comment]
>huh? I'm pretty sure the times I've intentionally used typeless loops were when I explicitly wanted to catch things that weren't the right type, not for the sake of optimization.

It came up a long time ago, when you were doing work on the process scheduler. It was recommend to use `:` for processing things in the huge lists for speed instead of `for(var/obj in processing_objects)` because of the istype checks---you found that by going with istypeless, it was faster than the above, and only a hair bit slower than using `:`.

**Comment by Krausus** [comment]
> you found that by going with istypeless, it was faster than the above, and only a hair bit slower than using `:`

That does ring a bell... but I ended up doing extra checks anyway, since it turned out that the performance of iterating over the list didn't really matter, but having processes not occasionally crash and burn did. Still, probably doesn't matter for most lists.

I don't remember the technique having an awful name like "istypeless", though.

**Comment by KasparoVy** [comment]
Antitypecast?

**Comment by Fox-McCloud** [comment]
If you reference Bay or Polaris (maybe VG) you can find the `trange` proc there (but if you re-add it, don't put it in explosion.dm)---that said, I'm really not sure if it's faster than `range` or not. Whichever is faster would be best.

**Comment by PJB3005** [comment]
Trange is definitely faster if you're only looking for turfs.

`range()` doesn't have the same byond internal optimization that `view()` does, where `for(var/x/y in view())` only passes `x` to the for loop.

`trange()` only ever returns turfs, so you're not looking through redundant things.

I actually profiled it and a view from a mob with all vision flags and such was only a tiny bit slower than trange.

**Comment by Plotron** [comment]
I can already feel much more depth with this!

It's as if the SS13 universe wasn't flat... Oh, wait, it is.

Anyway, loving the feel.

File: code/_onclick/hud/parallax.dm [hunk 5]
```diff
@@ -0,0 +1,288 @@
+/*
+ * This file handles all parallax-related business once the parallax itself is initialized with the rest of the HUD
+ */
+
+var/list/parallax_on_clients = list()
+
+/obj/screen/parallax
+	var/base_offset_x = 0
+	var/base_offset_y = 0
+	mouse_opacity = 1
```
Comment (clusterfack) [inline code/_onclick/hud/parallax.dm]: Suggest putting this back to 0

**Comment by clusterfack** [comment]
I considered using transform() for a while to manage this system, but didn't bother because I was only really trying to optimize this for the clientside and get it working, not to make it 100% performant like I normally would have. 

My suggestion for you is to try handling transform and animates by separating each layer into distinct planes and then using a transform matrix on each plane master instead of each object. Using animate in this way, *I believe* will be extremely costly.

File: code/_onclick/hud/parallax.dm [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+		break
+
+	if(forcerecalibrate || !parallax_loaded)
+		for(var/obj/screen/parallax/bgobj in C.parallax)
+			C.screen |= bgobj
+
+		C.screen |= C.parallax_master
+		C.screen |= C.parallax_spacemaster
+		C.screen |= C.parallax_dustmaster
+		C.parallax_dustmaster.color = list(0,0,0,0)
```
Comment (clusterfack) [inline code/_onclick/hud/parallax.dm]: A current bug with this is that the dustmaster shouldn't be taken out of the screen so that this has an effect

**Comment by Fox-McCloud** [comment]
Right so, I would really recommend just yanking out the parallax smoothing that you added in. 

I'm sure it really does look nice, but `animate` is nasty. If we wanted to look into adding smoothing later, with a more efficient methodology, then we could, but for now, I'd say this is a no-go.

Also, don't forget to re-implement `trange`.

**Comment by Fox-McCloud** [comment]
@monster860

Plan on working on this or not?

**Comment by monster860** [comment]
WTF.... Travis, stop giving me the finger already!

**Comment by Krausus** [comment]
Travis **fails** on warnings. You're getting warnings because you commented out the code that uses `acc_x/y_diff`, but not the definitions, so they're unused variables. And the warnings point to the exact lines that are causing them.

**Comment by Fox-McCloud** [comment]
Just revert the contents of this commit: https://github.com/ParadiseSS13/Paradise/pull/4739/commits/399a700285ede1961c5f4d76b65a959cbfab6366

And again re-implement `trange` as a helper/misc proc because of how speedy it is.

**Comment by monster860** [comment]
Got it.

File: code/_onclick/hud/parallax.dm [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+	C.previous_turf = posobj
+	C.last_parallax_shift = world_time
+
+	for(var/obj/screen/parallax/bgobj in C.parallax)
+		if(bgobj.parallax_speed)//only the middle and front layers actually move
+			var/accumulated_offset_x = bgobj.base_offset_x - round(C.parallax_offset["horizontal"] * bgobj.parallax_speed * (C.prefs.parallax_speed/2))
+			var/accumulated_offset_y = bgobj.base_offset_y - round(C.parallax_offset["vertical"] * bgobj.parallax_speed * (C.prefs.parallax_speed/2))
+
+			//var/acc_x_diff = bgobj.last_accumulated_offset_x - accumulated_offset_x
```
Comment (tigercat2000) [inline code/_onclick/hud/parallax.dm]: don't add commented out code

File: code/_onclick/hud/parallax.dm [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+				accumulated_offset_x += 1440
+
+			while(accumulated_offset_y > 720)
+				accumulated_offset_y -= 1440
+			while(accumulated_offset_y < -720)
+				accumulated_offset_y += 1440
+
+			bgobj.screen_loc = "CENTER-7:[accumulated_offset_x],CENTER-7:[accumulated_offset_y]"
+			
+			//if(abs(x_diff) <= 1 && abs(y_diff) <= 1)
```
Comment (tigercat2000) [inline code/_onclick/hud/parallax.dm]: don't add commented out code

Comment (monster860) [inline code/_onclick/hud/parallax.dm]: What if you want to add in smoothing later?

Comment (tigercat2000) [inline code/_onclick/hud/parallax.dm]: we won't
and if we do, git is commit based for a reason, can view old commits

**Comment by KasparoVy** [comment]
Pretty sure the conflict is with #4417 because of the SQL

**Comment by Signmark** [comment]
@monster860 muh conflicts

**Comment by monster860** [comment]
Morgue conflicts & lien notes fixed.

**Comment by KasparoVy** [comment]
:pray:

**Comment by Vivalas** [comment]
This is so beautiful.

**Comment by Regen1** [comment]
@TheDZD @Fox-McCloud 
Getting tiresome waiting for this to be merged
Last time it was because it has sql changes, but that was two patches ago

**Comment by TheDZD** [comment]
Conflicts yet again.

**Comment by Spacemanspark** [comment]
FINALLY.

**Comment by PJB3005** [comment]
:pray:

**Comment by Norgad** [comment]
💖💖🎇✨🎇✨🎇✨✨🎇🎆🎉🎉🎉✨🎊🎊✨❤❤💖

**Pull request merged**